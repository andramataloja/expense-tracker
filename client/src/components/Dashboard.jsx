import React, { Fragment } from "react"
import DoughnutChart from "./DoughnutChart"
import Container from '@material-ui/core/Container'
import Grid from '@material-ui/core/Grid';
import { makeStyles } from '@material-ui/core/styles';
import { Link } from "react-router-dom";
import {AddCircleOutline} from '@material-ui/icons';
import ListIcon from '@material-ui/icons/List';
import MonthSlider from './MonthSlider'

const useStyles = makeStyles({
    marginForSelection: {
      marginRight:'50px',
    },
    gridLayout: {
        marginTop:'45px',
        marginBottom:'45px'
      }
  });

export default function Dashboard() {  
    const classes = useStyles();
    return (
    <Fragment>
        <Container>
        <h1>Monthly Spending</h1>
        <MonthSlider/>
        <DoughnutChart/>       
        <Link to= '/addexpense'>
            <AddCircleOutline fontSize="large"/>
        </Link>              
        <Link to= '/allexpenses'>
            <ListIcon fontSize="large"/>
        </Link>
       
       {/*  <Grid container direction="column"
                justify="center"
                alignItems="center">
            <Grid item >
                <h1>Monthly Spending</h1>
            </Grid>
                <Grid item> <MonthSlider/></Grid>
            <Grid className={classes.gridLayout}> 
                <DoughnutChart/>
            </Grid>
            <Grid container justify="center" alignItems="center">
                <Grid className={classes.marginForSelection}>
                   <Link to= '/addexpense'>
                        <AddCircleOutline fontSize="large"/>
                    </Link>
                </Grid>
                <Grid item> 
                <Link to= '/allexpenses'>
                    <ListIcon fontSize="large"/>
                </Link>
                
                </Grid>
            </Grid>
        </Grid> */}
        </Container>
    </Fragment>
    )
  }
  