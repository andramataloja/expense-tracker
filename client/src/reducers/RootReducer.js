import * as types from "../actions/action-types";

export const initialState = {
  month: new Date().getMonth(),
  year: new Date().getFullYear(),
  doughnutData: [],
  expenses: [],
  index: ""
};

export const rootReducer = (state = initialState, action) => {
  switch (action.type) {
    case types.FETCH_MONTH:
      return {
        ...state,
        month: action.payload
      };
    case types.FETCH_YEAR:
      return {
        ...state,
        year: action.payload
      };
    case types.FETCH_DOUGHNUTDATA:
      return {
        ...state,
        doughnutData: action.payload
      };
    case types.FETCH_EXPENSES:
      return {
        ...state,
        expenses: action.payload
      };
    case types.SET_EXPLODEINDEX:
      return {
        ...state,
        index: action.payload
      };
    default:
      return state;
  }
};
export default rootReducer;
