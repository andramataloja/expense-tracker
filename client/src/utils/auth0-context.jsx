import React, { useState, useEffect, useContext } from "react";
import createAuth0Client from "@auth0/auth0-spa-js";

const DEFAULT_REDIRECT_CALLBACK = () =>
  window.history.replaceState({}, document.title, window.location.pathname);

export const Auth0Context = React.createContext();
export const useAuth0 = () => useContext(Auth0Context);
export const Auth0Provider = ({
  children,
  onRedirectCallback = DEFAULT_REDIRECT_CALLBACK
}) => {
  const [isAuthenticated, setIsAuthenticated] = useState();
  const [user, setUser] = useState();
  const [auth0Client, setAuth0Client] = useState();
  const [isLoading, setIsLoading] = useState(true);

  const config = {
    domain: process.env.REACT_APP_AUTH0_DOMAIN,
    client_id: process.env.REACT_APP_AUTH0_CLIENT_ID,
    redirect_uri: "http://localhost:3000/Dashboard"
  };

  useEffect(() => {
    const initializeAuth0 = async () => {
      const auth0FromHook = await createAuth0Client(config);
      setAuth0Client(auth0FromHook);

      if (
        window.location.search.includes("code=") &&
        window.location.search.includes("state=")
      ) {
        const { appState } = await auth0FromHook.handleRedirectCallback();
        onRedirectCallback(appState);
      }

      const isAuthenticated = await auth0FromHook.isAuthenticated();

      setIsAuthenticated(isAuthenticated);

      if (isAuthenticated) {
        const user = await auth0FromHook.getUser();
        setUser(user);
      }

      setIsLoading(false);
    };
    initializeAuth0();
  }, []);

  const handleRedirectCallback = async () => {
    setIsLoading(true);
    await auth0Client.handleRedirectCallback();
    const user = await auth0Client.getUser();
    setIsLoading(false);
    setIsAuthenticated(true);
    setUser(user);
  };
  return (
    <Auth0Context.Provider
      value={{
        isAuthenticated,
        user,
        isLoading,
        handleRedirectCallback,
        getIdTokenClaims: (...p) => auth0Client.getIdTokenClaims(...p),
        loginWithRedirect: (...p) => auth0Client.loginWithRedirect(...p),
        getTokenSilently: (...p) => auth0Client.getTokenSilently(...p),
        logout: (...p) => auth0Client.logout(...p)
      }}
    >
      {children}
    </Auth0Context.Provider>
  );
};
