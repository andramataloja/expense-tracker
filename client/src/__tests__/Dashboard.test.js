import React from "react";
import { cleanup, render, screen, wait } from "@testing-library/react";
import Dashboard from "../components/Dashboard";
import { useAuth0 } from "../utils/auth0-context";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";
import { initialState } from "../reducers/RootReducer";
import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import mockAxios from "axios";

const user = {
  email: "johndoe@me.com",
  email_verified: true,
  sub: "google-oauth2|12345678901234"
};

jest.mock("../utils/auth0-context");
jest.mock("../store.js");

describe("Dashboard", () => {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);

  afterEach(cleanup);
  beforeEach(() => {
    useAuth0.mockReturnValue({
      isAuthenticated: true,
      user,
      logout: jest.fn(),
      loginWithRedirect: jest.fn()
    });
    require("mutationobserver-shim");
    mockAxios.get = jest.fn(() => Promise.resolve(mockData));
  });

  const mockData = {
    data: [
      {
        category_id: 3,
        category_name: "Entertainment",
        icon: "Entertainment",
        fill: "#EA6E6E"
      },
      {
        category_id: 1,
        category_name: "Food",
        icon: "Food",
        fill: "#FCD246"
      }
    ]
  };

  const mockDough = {
    data: [
      {
        category_name: "Entertainment",
        total: "65",
        fill: "#EA6E6E"
      },
      {
        category_name: "Food",
        total: "78.5",
        fill: "#FCD246"
      }
    ]
  };

  it("should take a snapshot", async () => {
    const store = mockStore({
      ...initialState,
      doughnutData: mockDough
    });

    const { asFragment } = render(
      <Provider store={store}>
        <Dashboard />
      </Provider>
    );
    await act(wait);
    expect(asFragment(<Dashboard />)).toMatchSnapshot();
  });

  it("should render", async () => {
    const store = mockStore({
      ...initialState,
      doughnutData: mockDough
    });
    render(
      <Provider store={store}>
        <Dashboard />
      </Provider>
    );
    await act(wait);
    expect(screen.getByText("Monthly Spending")).toBeVisible();
    expect(screen.getByTestId("date-carousel")).toBeVisible();
    expect(screen.getByTestId("doughnutchart")).toBeVisible();
    expect(screen.getByTestId("add-button")).toBeVisible();
    expect(screen.getByTestId("expenses-paper")).toBeVisible();
  });
});
