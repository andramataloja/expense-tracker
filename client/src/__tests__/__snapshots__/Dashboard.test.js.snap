// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-2"
  >
    <div
      class="MuiBox-root MuiBox-root-3"
    >
      <div
        class="MuiBox-root MuiBox-root-6"
        data-testid="bar"
      >
        <header
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
          >
            <h5
              class="MuiTypography-root makeStyles-title-4 MuiTypography-h5"
            >
              Monthly Spending
            </h5>
            <div
              class="MuiBox-root MuiBox-root-78"
            >
              <div
                class="MuiBox-root MuiBox-root-79 makeStyles-username-5"
              />
              <button
                aria-controls="menu-appbar"
                aria-haspopup="true"
                aria-label="account of current user"
                class="MuiButtonBase-root MuiIconButton-root makeStyles-username-5"
                data-testid="open-menu"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </header>
      </div>
    </div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg"
    >
      <div
        class="MuiBox-root MuiBox-root-113"
      >
        <div
          class="MuiBox-root MuiBox-root-117 makeStyles-carouselSlides-116"
          data-testid="date-carousel"
        >
          <div>
            <div
              class="BrainhubCarousel makeStyles-carousel-114"
            >
              <div
                class="BrainhubCarousel__customArrows BrainhubCarousel__custom-arrowLeft"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-customArrow-115"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.62 2.99c-.49-.49-1.28-.49-1.77 0L6.54 11.3c-.39.39-.39 1.02 0 1.41l8.31 8.31c.49.49 1.28.49 1.77 0s.49-1.28 0-1.77L9.38 12l7.25-7.25c.48-.48.48-1.28-.01-1.76z"
                  />
                </svg>
              </div>
              <div
                class="BrainhubCarousel__trackContainer"
              >
                <ul
                  class="BrainhubCarousel__track BrainhubCarousel__track--transition BrainhubCarousel__track--draggable"
                  style="width: 0px; transform: translateX(-180px); margin-left: 0px; transition-duration: 500ms, 500ms;"
                >
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-197"
                    >
                      2019
                    </div>
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable BrainhubCarouselItem--active"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-198"
                    >
                      2020
                    </div>
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-199"
                    >
                      2021
                    </div>
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-200"
                    >
                      2022
                    </div>
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-201"
                    >
                      2023
                    </div>
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-202"
                    >
                      2024
                    </div>
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-203"
                    >
                      2025
                    </div>
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-204"
                    >
                      2026
                    </div>
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-205"
                    >
                      2027
                    </div>
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-206"
                    >
                      2028
                    </div>
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-207"
                    >
                      2029
                    </div>
                  </li>
                </ul>
              </div>
              <div
                class="BrainhubCarousel__customArrows BrainhubCarousel__custom-arrowRight"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-customArrow-115"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.38 21.01c.49.49 1.28.49 1.77 0l8.31-8.31c.39-.39.39-1.02 0-1.41L9.15 2.98c-.49-.49-1.28-.49-1.77 0s-.49 1.28 0 1.77L14.62 12l-7.25 7.25c-.48.48-.48 1.28.01 1.76z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div>
            <div
              class="BrainhubCarousel makeStyles-carousel-114"
            >
              <div
                class="BrainhubCarousel__customArrows BrainhubCarousel__custom-arrowLeft"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-customArrow-115"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.62 2.99c-.49-.49-1.28-.49-1.77 0L6.54 11.3c-.39.39-.39 1.02 0 1.41l8.31 8.31c.49.49 1.28.49 1.77 0s.49-1.28 0-1.77L9.38 12l7.25-7.25c.48-.48.48-1.28-.01-1.76z"
                  />
                </svg>
              </div>
              <div
                class="BrainhubCarousel__trackContainer"
              >
                <ul
                  class="BrainhubCarousel__track BrainhubCarousel__track--draggable"
                  style="width: 0px; margin-left: 0px; transform: translateX(-180px);"
                >
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-208"
                    />
                  </li>
                  <li
                    class="BrainhubCarouselItem BrainhubCarouselItem--clickable BrainhubCarouselItem--active"
                    style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
                  >
                    <div
                      class="MuiBox-root MuiBox-root-209"
                    />
                  </li>
                </ul>
              </div>
              <div
                class="BrainhubCarousel__customArrows BrainhubCarousel__custom-arrowRight"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-customArrow-115"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7.38 21.01c.49.49 1.28.49 1.77 0l8.31-8.31c.39-.39.39-1.02 0-1.41L9.15 2.98c-.49-.49-1.28-.49-1.77 0s-.49 1.28 0 1.77L14.62 12l-7.25 7.25c-.48.48-.48 1.28.01 1.76z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-120"
      >
        <div
          class="MuiBox-root MuiBox-root-121"
          data-testid="doughnutchart"
        />
      </div>
      <div
        class="MuiBox-root MuiBox-root-122"
      >
        <div
          class="MuiBox-root MuiBox-root-126"
        >
          <button
            class="MuiButtonBase-root MuiFab-root MuiFab-primary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiFab-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root makeStyles-addButton-123 MuiSvgIcon-fontSizeLarge"
                data-testid="add-button"
                focusable="false"
                role="button"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-152"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        data-testid="expenses-paper"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
